mm2src/coins/nft/trading_proto_v2/mod.rs

```diff
@@ -0,0 +1,549 @@
+            self.status_and_htlc_params_from_tx_data(
+                etomic_swap_contract,
+                &NFT_SWAP_CONTRACT,
+                &decoded,
+                index_bytes,
+                StateType::MakerPayments,
+            )
+            .await
+        );
+        match self.coin_type {
+            EthCoinType::Nft { .. } => {
+                let data =
+                    try_tx_s!(self.prepare_spend_nft_maker_v2_data(contract_type, &args, decoded, htlc_params, state));
+
+                self.sign_and_send_transaction(0.into(), Action::Call(etomic_swap_contract), data, U256::from(ETH_GAS))
+                    .compat()
+                    .await
+            },
+            EthCoinType::Eth | EthCoinType::Erc20 { .. } => Err(TransactionErr::ProtocolNotSupported(
+                "ETH and ERC20 Protocols are not supported for NFT Swaps".to_string(),
+            )),
+        }
+    }
+
+    pub(crate) async fn refund_nft_maker_payment_v2_timelock_impl(
+        &self,
+        args: RefundPaymentArgs<'_>,
+    ) -> Result<SignedEthTx, TransactionErr> {
+        let _etomic_swap_contract = try_tx_s!(args.swap_contract_address.try_to_address());
+        let tx: UnverifiedTransaction = try_tx_s!(rlp::decode(args.payment_tx));
+        let _payment = try_tx_s!(SignedEthTx::new(tx));
+        todo!()
+    }
+
+    fn prepare_nft_maker_payment_v2_data(
+        &self,
+        contract_type: &ContractType,
+        swap_contract_address: Address,
+        token_id: U256,
+        args: &SendNftMakerPaymentArgs<'_, Self>,
+        htlc_data: Vec<u8>,
+    ) -> Result<Vec<u8>, PrepareTxDataError> {
+        match contract_type {
+            ContractType::Erc1155 => {
+                let function = ERC1155_CONTRACT.function("safeTransferFrom")?;
+                let amount_u256 = U256::from_dec_str(&args.amount.to_string())
+                    .map_err(|e| PrepareTxDataError::Internal(e.to_string()))?;
+                let data = function.encode_input(&[
+                    Token::Address(self.my_address),
+                    Token::Address(swap_contract_address),
+                    Token::Uint(token_id),
+                    Token::Uint(amount_u256),
+                    Token::Bytes(htlc_data),
+                ])?;
+                Ok(data)
+            },
+            ContractType::Erc721 => {
+                let function = self.erc721_transfer_with_data()?;
+                let data = function.encode_input(&[
+                    Token::Address(self.my_address),
+                    Token::Address(swap_contract_address),
+                    Token::Uint(token_id),
+                    Token::Bytes(htlc_data),
+                ])?;
+                Ok(data)
+            },
+        }
+    }
+
+    fn validate_payment_args<'a>(
+        &self,
+        taker_secret_hash: &'a [u8],
+        maker_secret_hash: &'a [u8],
+        amount: &BigDecimal,
+        contract_type: &ContractType,
+    ) -> Result<(), String> {
```

Comment:
This doesn't need self param, also all params are retrieved from args. Should this be implemented for `EthCoin`?

Reply:
yeah, may be it is worth moving this function from Eth for future coins, so I removed self.
As necessary params are retrieved from different structures, I decided to "play" with generics https://github.com/KomodoPlatform/komodo-defi-framework/pull/2084/commits/8b5476b8d55aeebfb5cf38aebf8757c9f75b0090
may be on the one hand its over complicated, on the other the first step of using generic for contract type

Reply:
UPD: decided to validate params in send payment method and complicated work for myself :D

Reply:
> Should this be implemented for EthCoin?

I actually meant to just make `validate_payment_args` a standalone non public function, I am betting you will find out you don't need it when working on swap state machine and will remove it :)

> may be on the one hand its over complicated

It does over complicate things at this stage, let's make it a standalone function at this stage, btw for this https://github.com/KomodoPlatform/komodo-defi-framework/blob/defb7bb6f27270c07ebf2a88505aaebce8f696c9/mm2src/coins/nft/trading_proto_v2/mod.rs#L571-L576
secret hash length being 32 is specific to some coins only, e.g. for utxo/utxo we use 20 bytes length.

> btw validate_from_to_and_maker_status, prepare_spend_nft_maker_v2_data, erc721_transfer_with_data also dont use self

Let's get them out of `EthCoin` implementation then.

I just noticed that you moved the `EthCoin` implementation to `coins/nft/trading_proto_v2/mod.rs` when I thought it was inside the `eth` directory, I just meant to move them to a separate file inside `eth` directory similar to this https://github.com/KomodoPlatform/komodo-defi-framework/blob/evm-hd-wallet/mm2src/coins/eth/eth_hd_wallet.rs maybe name the file `nft_swap_v2`, when adding NFT support for other coins we add a similar file to the coin's dir. Let's leave in `coins/nft/trading_proto_v2/mod.rs` only implementations that will be used across all coins (if there are any, e.g. `validate_from_to_and_maker_status`, `prepare_spend_nft_maker_v2_data`, `erc721_transfer_with_data` and `validate_payment_args` are all curently related to eth, they should be non-public functions inside `nft_swap_v2` file in the `eth` dir).

Hope that is clear and better :)

Reply:
> I actually meant to just make validate_payment_args a standalone non public function, I am betting you will find out you don't need it when working on swap state machine and will remove it :)

:D oke oke, we will see. Just we have `validate_maker_payment_v2`, `validate_taker_payment_v2` ops in swap_v2 traits, so for sure we should have `validate_nft_maker_payment_v2` method. And Ive noticed we do the hash secret validation in legacy `send paymen`t and `validate payment` methods. Yeah I know it doesnt utilize state machine, but I decided it is worth adding such params check on this stage. For sure during state machine implementation for eth/nft if I see that there is no need in these params checks in send payment V2 and validate V2 functions, I will remove them.

> It does over complicate things at this stage, let's make it a standalone function at this stage,

Yeah, I made it standalone, just used generics for args structures. You are right, as actually this functionality is related only to eth, there is no need in over complicating  things. Removed generics.

simplified validate function https://github.com/KomodoPlatform/komodo-defi-framework/pull/2084/commits/71db8d77cdfd02b73d21bb4124bd7f28a1f8f1de and made functions which dont need self standalone https://github.com/KomodoPlatform/komodo-defi-framework/pull/2084/commits/71f4df5076c911f085b6a3d50d8f22320b261e18

> I just noticed that you moved the EthCoin implementation to coins/nft/trading_proto_v2/mod.rs when I thought it was inside the eth directory, I just meant to move them to a separate file inside eth directory

Yep, sorry I so wanted to keep `nft trading` and future `eth trading` implementations separately, that I missed the part that actually right now we dont have common nft types for other coins, and all structs (except args, but we keep them in lp_coins.rs) and errors are related to EthCoin.
You are absolutely right, that we need to move eth nft impl to eth crate. As there are already 4 error types for Eth nft swaps I decided to keep it as `mod` instead of just file. I think it would be cleaner, as may be there will be more specific errors for refund methods. But renamed module to `nft_swap_v2`. So we have `eth/nft_swap_v2/mod.rs`.
For trading part may be we will have `eth/trading_proto_v2/mod.rs` or `eth/eth_swap_v2/mod.rs`, will see

fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2084/commits/8e6183cfd2e1723d4b53e82554ccd52787b7d121

Reply:
> :D oke oke, we will see. Just we have validate_maker_payment_v2, validate_taker_payment_v2 ops in swap_v2 traits, so for sure we should have validate_nft_maker_payment_v2 method.

I meant `validate_payment_args` especially at the maker's side, we need `validate_nft_maker_payment_v2` for sure on the taker's side. I guess we will actually need `validate_payment_args` but it will not be reused anywhere else other than `validate_nft_maker_payment_v2`, for `validate_nft_taker_funding`, `maker_secret_hash` will not be available yet and we will not be able to reuse this functions. Anyway, we will wait and see :)